File: examples/typescript/heapExamples.ts
class MinHeap<T> {
    private heap: T[] = [];
    constructor(private compare: (a: T, b: T) => number) {}
    
    push(val: T): void {
        this.heap.push(val);
        this.bubbleUp(this.heap.length - 1);
    }
    
    pop(): T | undefined {
        if (this.heap.length === 0) return undefined;
        if (this.heap.length === 1) return this.heap.pop();
        
        const top = this.heap[0];
        this.heap[0] = this.heap.pop()!;
        this.bubbleDown(0);
        return top;
    }
    
    peek(): T | undefined {
        return this.heap[0];
    }
    
    size(): number {
        return this.heap.length;
    }
    
    private bubbleUp(index: number): void {
        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            if (this.compare(this.heap[index], this.heap[parentIndex]) >= 0) break;
            [this.heap[index], this.heap[parentIndex]] = [this.heap[parentIndex], this.heap[index]];
            index = parentIndex;
        }
    }
    
    private bubbleDown(index: number): void {
        while (true) {
            let minIndex = index;
            const leftChild = 2 * index + 1;
            const rightChild = 2 * index + 2;
            
            if (leftChild < this.heap.length && this.compare(this.heap[leftChild], this.heap[minIndex]) < 0) {
                minIndex = leftChild;
            }
            if (rightChild < this.heap.length && this.compare(this.heap[rightChild], this.heap[minIndex]) < 0) {
                minIndex = rightChild;
            }
            if (minIndex === index) break;
            
            [this.heap[index], this.heap[minIndex]] = [this.heap[minIndex], this.heap[index]];
            index = minIndex;
        }
    }
}

// 1. Kth Largest Element in an Array
function findKthLargest(nums: number[], k: number): number {
    const minHeap = new MinHeap<number>((a, b) => a - b);
    for (const num of nums) {
        minHeap.push(num);
        if (minHeap.size() > k) minHeap.pop();
    }
    return minHeap.peek()!;
}

// 2. Top K Frequent Elements
function topKFrequent(nums: number[], k: number): number[] {
    const count = new Map<number, number>();
    for (const num of nums) {
        count.set(num, (count.get(num) || 0) + 1);
    }
    
    const minHeap = new MinHeap<[number, number]>((a, b) => a[1] - b[1]);
    for (const [num, freq] of count) {
        minHeap.push([num, freq]);
        if (minHeap.size() > k) minHeap.pop();
    }
    
    const result: number[] = [];
    while (minHeap.size() > 0) {
        result.push(minHeap.pop()![0]);
    }
    return result;
}

// 3. Merge k Sorted Lists
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = val === undefined ? 0 : val;
        this.next = next === undefined ? null : next;
    }
}

function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
    const minHeap = new MinHeap<[number, number, ListNode]>((a, b) => a[0] - b[0]);
    
    for (let i = 0; i < lists.length; i++) {
        if (lists[i]) {
            minHeap.push([lists[i]!.val, i, lists[i]!]);
        }
    }
    
    const dummy = new ListNode(0);
    let curr = dummy;
    
    while (minHeap.size() > 0) {
        const [val, i, node] = minHeap.pop()!;
        curr.next = node;
        curr = curr.next;
        
        if (node.next) {
            minHeap.push([node.next.val, i, node.next]);
        }
    }
    
    return dummy.next;
}

// 4. Find Median from Data Stream
class MedianFinder {
    private maxHeap = new MinHeap<number>((a, b) => b - a); // for smaller half
    private minHeap = new MinHeap<number>((a, b) => a - b); // for larger half
    
    addNum(num: number): void {
        this.maxHeap.push(num);
        this.minHeap.push(this.maxHeap.pop()!);
        
        if (this.minHeap.size() > this.maxHeap.size()) {
            this.maxHeap.push(this.minHeap.pop()!);
        }
    }
    
    findMedian(): number {
        return this.maxHeap.size() > this.minHeap.size() 
            ? this.maxHeap.peek()! 
            : (this.maxHeap.peek()! + this.minHeap.peek()!) / 2;
    }
}

// 5. K Closest Points to Origin
function kClosest(points: number[][], k: number): number[][] {
    const maxHeap = new MinHeap<[number, number[]]>((a, b) => b[0] - a[0]);
    
    for (const point of points) {
        const dist = point[0] * point[0] + point[1] * point[1];
        maxHeap.push([dist, point]);
        if (maxHeap.size() > k) maxHeap.pop();
    }
    
    const result: number[][] = [];
    while (maxHeap.size() > 0) {
        result.push(maxHeap.pop()![1]);
    }
    return result;
}

// 6. Last Stone Weight
function lastStoneWeight(stones: number[]): number {
    const maxHeap = new MinHeap<number>((a, b) => b - a);
    for (const stone of stones) maxHeap.push(stone);
    
    while (maxHeap.size() > 1) {
        const first = maxHeap.pop()!;
        const second = maxHeap.pop()!;
        if (first !== second) {
            maxHeap.push(first - second);
        }
    }
    
    return maxHeap.size() === 0 ? 0 : maxHeap.peek()!;
}

// 7. Task Scheduler
function leastInterval(tasks: string[], n: number): number {
    const count = new Map<string, number>();
    for (const task of tasks) {
        count.set(task, (count.get(task) || 0) + 1);
    }
    
    const maxHeap = new MinHeap<number>((a, b) => b - a);
    for (const freq of count.values()) {
        if (freq > 0) maxHeap.push(freq);
    }
    
    let time = 0;
    while (maxHeap.size() > 0) {
        const temp: number[] = [];
        for (let i = 0; i <= n; i++) {
            if (maxHeap.size() > 0) {
                const freq = maxHeap.pop()!;
                if (freq > 1) temp.push(freq - 1);
            }
            time++;
            if (maxHeap.size() === 0 && temp.length === 0) break;
        }
        for (const freq of temp) {
            maxHeap.push(freq);
        }
    }
    return time;
}

// 8. Ugly Number II
function nthUglyNumber(n: number): number {
    const minHeap = new MinHeap<number>((a, b) => a - b);
    const seen = new Set<number>();
    
    minHeap.push(1);
    seen.add(1);
    
    let ugly = 1;
    for (let i = 0; i < n; i++) {
        ugly = minHeap.pop()!;
        for (const factor of [2, 3, 5]) {
            const next = ugly * factor;
            if (!seen.has(next)) {
                minHeap.push(next);
                seen.add(next);
            }
        }
    }
    return ugly;
}

// 9. Meeting Rooms II
function minMeetingRooms(intervals: number[][]): number {
    intervals.sort((a, b) => a[0] - b[0]);
    const minHeap = new MinHeap<number>((a, b) => a - b);
    
    for (const [start, end] of intervals) {
        if (minHeap.size() > 0 && minHeap.peek()! <= start) {
            minHeap.pop();
        }
        minHeap.push(end);
    }
    return minHeap.size();
}

// 10. Kth Smallest Element in a Sorted Matrix
function kthSmallest(matrix: number[][], k: number): number {
    const minHeap = new MinHeap<[number, number, number]>((a, b) => a[0] - b[0]);
    
    for (let i = 0; i < matrix.length; i++) {
        minHeap.push([matrix[i][0], i, 0]);
    }
    
    for (let i = 0; i < k - 1; i++) {
        const [val, row, col] = minHeap.pop()!;
        if (col + 1 < matrix[0].length) {
            minHeap.push([matrix[row][col + 1], row, col + 1]);
        }
    }
    return minHeap.peek()![0];
}

// Test cases
console.log("=== Heap Examples ===");

// Test 1: Kth Largest Element
console.log("Kth Largest([3,2,1,5,6,4], k=2):", findKthLargest([3,2,1,5,6,4], 2));

// Test 2: Top K Frequent Elements
console.log("Top K Frequent([1,1,1,2,2,3], k=2):", topKFrequent([1,1,1,2,2,3], 2));

// Test 4: Median Finder
const mf = new MedianFinder();
mf.addNum(1);
mf.addNum(2);
console.log("Median after adding 1,2:", mf.findMedian());

// Test 5: K Closest Points
const points = [[1,3],[-2,2]];
console.log("K Closest Points(k=1):", kClosest(points, 1));

// Test 6: Last Stone Weight
console.log("Last Stone Weight([2,7,4,1,8,1]):", lastStoneWeight([2,7,4,1,8,1]));

// Test 7: Task Scheduler
console.log("Least Interval(['A','A','A','B','B','B'], n=2):", leastInterval(['A','A','A','B','B','B'], 2));

// Test 8: Ugly Number II
console.log("10th Ugly Number:", nthUglyNumber(10));

// Test 9: Meeting Rooms II
const meetings = [[0,30],[5,10],[15,20]];
console.log("Min Meeting Rooms:", minMeetingRooms(meetings));

// Test 10: Kth Smallest in Matrix
const matrix = [[1,5,9],[10,11,13],[12,13,15]];
console.log("Kth Smallest in Matrix(k=8):", kthSmallest(matrix, 8));

File: examples/python/heap_examples.py
import heapq
from collections import defaultdict, Counter

# 1. Find Kth Largest
def find_kth_largest(nums, k):
    heap = []
    for num in nums:
        heapq.heappush(heap, num)
        if len(heap) > k:
            heapq.heappop(heap)
    return heap[0]

# 2. Top K Frequent
def top_k_frequent(nums, k):
    count = Counter(nums)
    heap = []
    for num, freq in count.items():
        heapq.heappush(heap, (freq, num))
        if len(heap) > k:
            heapq.heappop(heap)
    return [num for freq, num in heap]

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# 3. Find Median Data Stream
def find_median_data_stream():
    mf = MedianFinder()
    mf.add_num(1)
    mf.add_num(2)
    return mf.find_median()

# 4. Merge K Lists
def merge_k_lists(lists):
    heap = []
    for i, node in enumerate(lists):
        if node:
            heapq.heappush(heap, (node.val, i, node))
    
    dummy = ListNode(0)
    curr = dummy
    while heap:
        val, i, node = heapq.heappop(heap)
        curr.next = node
        curr = curr.next
        if node.next:
            heapq.heappush(heap, (node.next.val, i, node.next))
    return dummy.next

class MedianFinder:
    def __init__(self):
        self.max_heap = []  # for smaller half
        self.min_heap = []  # for larger half
    
    def add_num(self, num):
        heapq.heappush(self.max_heap, -num)
        heapq.heappush(self.min_heap, -heapq.heappop(self.max_heap))
        if len(self.min_heap) > len(self.max_heap):
            heapq.heappush(self.max_heap, -heapq.heappop(self.min_heap))
    
    def find_median(self):
        if len(self.max_heap) > len(self.min_heap):
            return -self.max_heap[0]
        return (-self.max_heap[0] + self.min_heap[0]) / 2.0

# 5. K Closest
def k_closest(points, k):
    heap = []
    for x, y in points:
        dist = -(x*x + y*y)  # negative for max heap
        heapq.heappush(heap, (dist, x, y))
        if len(heap) > k:
            heapq.heappop(heap)
    return [[x, y] for dist, x, y in heap]

# 6. Kth Smallest
def kth_smallest(matrix, k):
    heap = []
    for i in range(len(matrix)):
        heapq.heappush(heap, (matrix[i][0], i, 0))
    
    for _ in range(k - 1):
        val, row, col = heapq.heappop(heap)
        if col + 1 < len(matrix[0]):
            heapq.heappush(heap, (matrix[row][col + 1], row, col + 1))
    return heap[0][0]

# 7. Last Stone Weight
def last_stone_weight(stones):
    heap = [-stone for stone in stones]
    heapq.heapify(heap)
    
    while len(heap) > 1:
        first = -heapq.heappop(heap)
        second = -heapq.heappop(heap)
        if first != second:
            heapq.heappush(heap, -(first - second))
    return -heap[0] if heap else 0

# 8. Least Interval
def least_interval(tasks, n):
    count = Counter(tasks)
    heap = [-freq for freq in count.values()]
    heapq.heapify(heap)
    
    time = 0
    while heap:
        temp = []
        for _ in range(n + 1):
            if heap:
                freq = heapq.heappop(heap)
                if freq < -1:
                    temp.append(freq + 1)
            time += 1
            if not heap and not temp:
                break
        for freq in temp:
            heapq.heappush(heap, freq)
    return time

# 9. Nth Ugly Number
def nth_ugly_number(n):
    heap = [1]
    seen = {1}
    
    for _ in range(n):
        ugly = heapq.heappop(heap)
        for factor in [2, 3, 5]:
            next_ugly = ugly * factor
            if next_ugly not in seen:
                heapq.heappush(heap, next_ugly)
                seen.add(next_ugly)
    return ugly

# 10. Min Meeting Rooms
def min_meeting_rooms(intervals):
    intervals.sort()
    heap = []
    
    for start, end in intervals:
        if heap and heap[0] <= start:
            heapq.heappop(heap)
        heapq.heappush(heap, end)
    return len(heap)

def run_examples():
    print("=== Heap Examples ===")
    
    # Test 1: Kth Largest Element
    print(f"Kth Largest([3,2,1,5,6,4], k=2): {find_kth_largest([3,2,1,5,6,4], 2)}")
    
    # Test 2: Top K Frequent Elements
    print(f"Top K Frequent([1,1,1,2,2,3], k=2): {top_k_frequent([1,1,1,2,2,3], 2)}")
    
    # Test 3: Find Median Data Stream
    print(f"Find Median Data Stream: {find_median_data_stream()}")
    
    # Test 4: Median Finder
    mf = MedianFinder()
    mf.add_num(1)
    mf.add_num(2)
    print(f"Median after adding 1,2: {mf.find_median()}")
    
    # Test 5: K Closest Points
    points = [[1,3],[-2,2]]
    print(f"K Closest Points(k=1): {k_closest(points, 1)}")
    
    # Test 6: Kth Smallest in Matrix
    matrix = [[1,5,9],[10,11,13],[12,13,15]]
    print(f"Kth Smallest in Matrix(k=8): {kth_smallest(matrix, 8)}")
    
    # Test 7: Last Stone Weight
    print(f"Last Stone Weight([2,7,4,1,8,1]): {last_stone_weight([2,7,4,1,8,1])}")
    
    # Test 8: Task Scheduler
    print(f"Least Interval(['A','A','A','B','B','B'], n=2): {least_interval(['A','A','A','B','B','B'], 2)}")
    
    # Test 9: Ugly Number II
    print(f"10th Ugly Number: {nth_ugly_number(10)}")
    
    # Test 10: Meeting Rooms II
    meetings = [[0,30],[5,10],[15,20]]
    print(f"Min Meeting Rooms: {min_meeting_rooms(meetings)}")

if __name__ == "__main__":
    run_examples()